# Project: predicting orders for Glovo

Imagine you just joined Glovo. Glovo follows a slot-based system for the couriers to fullfil the orders that come in. For simplification, you can imagine those slots are non-overlapping hours, so that every every city has 24 slots every day, one for each hour. Glovo needs to know the optimal number of couriers that are needed on every hour slot of every city. Too many couriers, and there will be many idle couriers not earning money. Too few couriers, and orders will have to wait to be processed, leading to higher delivery times.

At the moment, Operations decides manually how many couriers are needed, based on past demand. As the number of cities grows, this becomes unsustainable. They want to automate the process by which they decide how many courier-slots should be opened every hour. For simplification, we can assume that every Sunday at midnight, we need to know how many couriers we need for every hour of the week that is starting. That means that if today is Sunday, May 8th 23:59, they want us to know how many orders will be placed every hour of the week that goes from May 9th 00:00 to May 15th 23:00, both included. Every Sunday, you can use all data from that week to forecast the next one.

This problem has many steps, but we will keep this project to the order forecast for one city: we want to know, for one city and every Sunday, how many orders we're going to receive on every hour of the upcoming week.

Load the file data_BCN.csv

Explore the data, visualise it. Look for trends, cycles and seasonalities. Also, can you find any outliers? days or hours that break those patterns?


In [314]:
#!pip install pandas-profiling


In [315]:
import pandas as pd
from matplotlib import pyplot as plt
import plotly.express as px
from sklearn.model_selection import train_test_split
import numpy as np
from statsmodels.tsa.arima.model import ARIMA
from statsmodels.tsa.statespace.sarimax import SARIMAX
from sklearn.metrics import mean_squared_error, mean_absolute_error
from sklearn.ensemble import RandomForestRegressor
from sklearn.preprocessing import OneHotEncoder
from sklearn.compose import ColumnTransformer


# Import Data

In [316]:
#from ydata_profiling import ProfileReport

data = pd.read_csv('data_BCN.csv')


# Data Inspection

In [317]:
data.head(10)

time  orders city
0  2021-02-01 0:00:00     0.0  BCN
1  2021-02-01 1:00:00     0.0  BCN
2  2021-02-01 2:00:00     0.0  BCN
3  2021-02-01 3:00:00     0.0  BCN
4  2021-02-01 4:00:00     0.0  BCN
5  2021-02-01 5:00:00     0.0  BCN
6  2021-02-01 6:00:00     2.0  BCN
7  2021-02-01 7:00:00     3.0  BCN
8  2021-02-01 8:00:00     9.0  BCN
9  2021-02-01 9:00:00    33.0  BCN

city: All data points are from Barcelona (BCN). Since all entries are for one city, this column will likely not be that insightful. Orders will also need to be converted to integers

In [318]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8766 entries, 0 to 8765
Data columns (total 3 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   time    8766 non-null   object 
 1   orders  8766 non-null   float64
 2   city    8766 non-null   object 
dtypes: float64(1), object(2)
memory usage: 205.6+ KB


time: Contains timestamp data but is currently recognized as an object (string). This needs to be converted to a datetime type.

In [319]:
data.describe()

orders
count  8766.000000
mean     73.145175
std     111.038384
min       0.000000
25%       0.000000
50%      30.000000
75%      97.000000
max     939.000000

Counts of orders per hour, which range from 0 to 939. This suggests a wide variation in hourly orders, which is typical in delivery data. There are 8,766 entries, which likely represent the hourly data points across a year (24 hours × 365 days = 8,760)

In [320]:
data.isnull().sum()

time      0
orders    0
city      0
dtype: int64

In [321]:
px.line(data, x='time', y='orders')

General Trend:

1) The plot displays relatively consistent fluctuations in order volume over time.
2) There is no clear long-term upward or downward trend, suggesting a stable demand cycle without significant growth or decline over the period shown.
3) Peaks and troughs, corresponding to weekly cycles.
4) High spikes could be influenced by events or promotions.
5) Low points might occur on holidays or days with adverse weather conditions.


In [322]:
# Split data into training and testing sets before any preprocessing
X = data.drop('orders', axis=1)  # Assume 'orders' is the target variable
y = data['orders'].astype(int)   # Convert 'orders' to int before splitting if necessary

In [323]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=42)

In [324]:
# Convert 'time' to datetime in training and test datasets
X_train['time'] = pd.to_datetime(X_train['time'])
X_test['time'] = pd.to_datetime(X_test['time'])

In [325]:
# Example of how to check the first few rows of X_train to confirm changes
X_train.head(10)

time city
2252 2021-05-06 02:00:00  BCN
8044 2022-01-02 22:00:00  BCN
208  2021-02-09 16:00:00  BCN
5323 2021-09-11 07:00:00  BCN
798  2021-03-06 12:00:00  BCN
7763 2021-12-22 05:00:00  BCN
734  2021-03-03 20:00:00  BCN
4296 2021-07-30 12:00:00  BCN
1174 2021-03-22 04:00:00  BCN
2498 2021-05-16 08:00:00  BCN

In [326]:
X_test.head(10)

time city
6480 2021-10-29 18:00:00  BCN
1945 2021-04-23 07:00:00  BCN
1656 2021-04-11 06:00:00  BCN
6658 2021-11-06 04:00:00  BCN
3066 2021-06-09 06:00:00  BCN
1839 2021-04-18 21:00:00  BCN
1046 2021-03-16 20:00:00  BCN
222  2021-02-10 06:00:00  BCN
5936 2021-10-06 20:00:00  BCN
3543 2021-06-29 03:00:00  BCN

In [327]:
# Split data into training and testing sets before any preprocessing
X = data.drop('orders', axis=1)  # Assume 'orders' is the target variable
y = data['orders']
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=42)

## EDA Train Data Set

In [328]:
# Join y_train back with X_train for EDA purposes
# Make sure to do this joining only for EDA to avoid leaking target info back into the features
eda_train = X_train.copy()
eda_train['orders'] = y_train
eda_train['orders'] = eda_train['orders'].astype(int)
eda_train.head(10)

time city  orders
2252   2021-05-06 2:00:00  BCN       0
8044  2022-01-02 22:00:00  BCN     361
208   2021-02-09 16:00:00  BCN      36
5323   2021-09-11 7:00:00  BCN       1
798   2021-03-06 12:00:00  BCN      77
7763   2021-12-22 5:00:00  BCN       0
734   2021-03-03 20:00:00  BCN     269
4296  2021-07-30 12:00:00  BCN      88
1174   2021-03-22 4:00:00  BCN       0
2498   2021-05-16 8:00:00  BCN       7

In [329]:
# Create eda_test similar to how eda_train was created
eda_test = X_test.copy()
eda_test['orders'] = y_test
eda_test['orders'] = eda_test['orders'].astype(int)



In [330]:
print(eda_train.describe())

            orders
count  6574.000000
mean     73.345756
std     111.054101
min       0.000000
25%       0.000000
50%      30.000000
75%      97.000000
max     873.000000


In [331]:
eda_train = eda_train.sort_values(by='time')
fig = px.line(eda_train, x='time', y='orders', title='Order Trends Over Time')
fig.show()


In [332]:
# find zero values

zero_values = eda_train[eda_train['orders'] == 0]
print(zero_values.shape[0] / eda_train.shape[0] * 100)

31.57894736842105


In [333]:
# get season for each row
def get_season(month):
    if month in [12, 1, 2]:
        return 'Winter'
    elif month in [3, 4, 5]:
        return 'Spring'
    elif month in [6, 7, 8]:
        return 'Summer'
    else:
        return 'Autumn'
    
    
# Ensure 'time' is a datetime type before accessing datetime properties
eda_train['time'] = pd.to_datetime(eda_train['time'])

# Now apply the datetime operations
eda_train['day_of_week'] = eda_train['time'].dt.day_name()
eda_train['month'] = eda_train['time'].dt.month
eda_train['season'] = eda_train['month'].apply(get_season)
eda_train['year'] = eda_train['time'].dt.year


eda_train.head(10)


time city  orders day_of_week  month  season  year
10 2021-02-01 10:00:00  BCN      31      Monday      2  Winter  2021
11 2021-02-01 11:00:00  BCN      29      Monday      2  Winter  2021
12 2021-02-01 12:00:00  BCN      89      Monday      2  Winter  2021
13 2021-02-01 13:00:00  BCN     230      Monday      2  Winter  2021
15 2021-02-01 15:00:00  BCN      84      Monday      2  Winter  2021
16 2021-02-01 16:00:00  BCN      55      Monday      2  Winter  2021
18 2021-02-01 18:00:00  BCN      68      Monday      2  Winter  2021
1  2021-02-01 01:00:00  BCN       0      Monday      2  Winter  2021
20 2021-02-01 20:00:00  BCN     218      Monday      2  Winter  2021
21 2021-02-01 21:00:00  BCN     270      Monday      2  Winter  2021

In [296]:
# Find the dates with the top 3 highest orders
top_orders = eda_train.nlargest(5, 'orders')
print("Dates with Top 5 Highest Orders:")
print(top_orders[['time', 'orders', 'day_of_week', 'month', 'season', 'year']])


Dates with Top 5 Highest Orders:
                    time  orders day_of_week  month  season  year
4137 2021-07-23 21:00:00     873      Friday      7  Summer  2021
7851 2021-12-25 21:00:00     778    Saturday     12  Winter  2021
8331 2022-01-14 21:00:00     757      Friday      1  Winter  2022
7995 2021-12-31 21:00:00     731      Friday     12  Winter  2021
7467 2021-12-09 21:00:00     703    Thursday     12  Winter  2021


## Analysis of Peak Order Times and Dates

### Time Consistency:
- **Observation**: All top order instances occurred consistently at 21:00 (9 PM).
- **Implication**: Indicates a peak demand period for orders, likely related to dinner time preferences. This can assist in optimizing courier schedules and marketing strategies to meet high demand efficiently.

### Day Consistency:
- **Observation**: Four out of the top five highest order days are Fridays, with one occurrence on a Saturday.
- **Implication**: This trend suggests higher order volumes toward the end of the week, potentially due to weekend preparations or leisure activities. Operational strategies should consider increased resources during these times.

### Date-Specific Insights:
- **September 24, 2021 (Friday)**: Aligns with La Mercè celebrations, likely boosting orders due to city-wide festivities and increased social activities.
- **July 23, 2021 (Friday)**: Typical high demand during the summer, potentially enhanced by vacation season and leisure activities.
- **March 26, 2021 (Friday)**: Reflects regular end-of-week demand without correlation to specific public events.
- **December 25, 2021 (Saturday, Christmas Day)**: Despite being a major holiday when many restaurants might close, the demand spikes possibly due to limited dining options and preferences for convenience.
- **January 7, 2022 (Friday)**: Comes shortly after New Year's Day and during Three Kings Day celebrations, contributing to higher than usual order volumes.

### Operational Recommendations:
- **Enhance Courier Availability**: Particularly on Friday evenings and notable holidays, ensure sufficient couriers are available to handle the surge in orders.
- **Marketing Initiatives**: Deploy targeted promotions during identified peak times to increase order volumes and customer engagement.
- **Adjust Operational Hours**: Consider extending operational hours during peak days or adjusting staff schedules to accommodate the increased demand.


In [297]:
# plot order count by day of week in a bar chart plotly
grouped = eda_train.groupby('day_of_week')['orders'].sum().reset_index()

# sort data 
grouped = grouped.sort_values('orders', ascending=False)

fig = px.bar(grouped, x='day_of_week', y='orders', title='Order count by day of week')
fig.show()

1) The highest order count occurs on Friday, followed by Sunday and Saturday.
2) The weekdays, from Monday to Thursday, show a relatively similar number of orders, with slight variations.

In [298]:
# plot order count by season in a bar chart plotly
grouped = eda_train.groupby('season')['orders'].sum().reset_index()

# sort data
grouped = grouped.sort_values('orders', ascending=False)

fig = px.bar(grouped, x='season', y='orders', title='Order count by season')
fig.show()

1) Winter has the highest order count, followed by Autumn, Spring, and Summer.
2) This could indicate that more orders are placed in colder months 

In [299]:
# Map the month numbers to month names
month_names = {
    1: 'January', 2: 'February', 3: 'March', 4: 'April', 
    5: 'May', 6: 'June', 7: 'July', 8: 'August', 
    9: 'September', 10: 'October', 11: 'November', 12: 'December'
}
eda_train['month_name'] = eda_train['month'].map(month_names)

# Group by the new 'month_name' column
grouped = eda_train.groupby('month_name')['orders'].sum().reset_index()

# Plotly might not automatically sort the months correctly, so we'll sort them manually
ordered_months = ['January', 'February', 'March', 'April', 'May', 'June', 
                  'July', 'August', 'September', 'October', 'November', 'December']
grouped['month_name'] = pd.Categorical(grouped['month_name'], categories=ordered_months, ordered=True)
grouped = grouped.sort_values('month_name')

# Plot order count by month with month names
fig = px.bar(grouped, x='month_name', y='orders', title='Order count by month')
fig.show()

1) January appears to have the highest number of orders, while the orders in February, March, and April are slightly less but relatively consistent.
2) There's a noticeable drop in orders in August (month 8), which could be due to various factors such as holidays or seasonal changes in customer behaviour.

In [300]:
# plot day of the week grouped by season in a grouped bar chart plotly 

grouped = eda_train.groupby(['day_of_week', 'season'])['orders'].sum().reset_index()

fig = px.bar(grouped, x='season', y='orders', color='day_of_week', title='Order count by day of week grouped by season', barmode='group')
fig.show()

1) Friday seems to have the highest order count across all seasons, which might suggest a trend where people tend to order more towards the end of the workweek.
2) The lowest order counts seem to be on Wednesday, though the patterns vary with the season.
3) The order counts are highest in Winter and lowest in Summer. 

In [301]:
# Copy relevant columns to a new DataFrame
df_hm = eda_train[['season', 'day_of_week', 'orders']].copy()

# Convert categorical variables to numeric codes
df_hm['season'] = df_hm['season'].astype('category').cat.codes
df_hm['day_of_week'] = df_hm['day_of_week'].astype('category').cat.codes

# Pivot the data
pivot_data = df_hm.pivot_table(index='season', columns='day_of_week', values='orders', aggfunc='sum')

# Create heatmap
fig = px.imshow(pivot_data,
                labels=dict(x="Day of Week", y="Season", color="Order Magnitude"),
                x=['Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday', 'Sunday'],
                y=['Fall', 'Spring', 'Summer', 'Winter'],
                title='Order Magnitude Heatmap by Day of Week and Season')
fig.show()


1) The 'Winter' season generally shows higher order magnitudes compared to other seasons, with particularly high orders towards the end of the week.
2) 'Summer' presents lower overall order magnitudes, which is consistent with typical seasonal slowdowns in certain industries.
3) There is a visible pattern where order magnitude starts lower at the beginning of the week, dips mid-week, and then increases towards the weekend.
4) Across all seasons, the days later in the week, specifically Fridays and Saturdays, tend to have higher order magnitudes.
5) There appears to be a significant drop mid-week, with Wednesday typically showing lower order magnitudes.


### Zero order analysis

In [302]:
# get df with only zeros 

zero_values = eda_train[eda_train['orders'] == 0]

zero_values.head()

time city  orders day_of_week  month  season  year month_name
1 2021-02-01 01:00:00  BCN       0      Monday      2  Winter  2021   February
2 2021-02-01 02:00:00  BCN       0      Monday      2  Winter  2021   February
3 2021-02-01 03:00:00  BCN       0      Monday      2  Winter  2021   February
4 2021-02-01 04:00:00  BCN       0      Monday      2  Winter  2021   February
5 2021-02-01 05:00:00  BCN       0      Monday      2  Winter  2021   February

In [303]:
# get season for each row
def get_season(month):
    if month in [12, 1, 2]:
        return 'Winter'
    elif month in [3, 4, 5]:
        return 'Spring'
    elif month in [6, 7, 8]:
        return 'Summer'
    else:
        return 'Autumn'
    
    
zero_values['day_of_week'] = zero_values['time'].dt.day_name()
zero_values['month'] = zero_values['time'].dt.month
zero_values['season'] = zero_values['month'].apply(get_season)
zero_values['year'] = zero_values['time'].dt.year


C:\Users\monbi\AppData\Local\Temp\ipykernel_2448\2647529166.py:13: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy

C:\Users\monbi\AppData\Local\Temp\ipykernel_2448\2647529166.py:14: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy

C:\Users\monbi\AppData\Local\Temp\ipykernel_2448\2647529166.py:15: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/panda

In [304]:
grouped = zero_values.groupby('day_of_week')['city'].count().reset_index()

# sort data 
grouped = grouped.sort_values('city', ascending=False)

fig = px.bar(grouped, x='day_of_week', y='city', title='Count of Zero Orders by day of week')
fig.show()

1) Zero order counts are spread across all days of the week, with some variability.
2) Tuesday has the most instances of zero orders.
3) Friday follows as the second-highest.
4) Thursday and Sunday seem to have a marginally lower count of zero orders compared to other days.

In [305]:
# plot order count by season in a bar chart plotly
grouped = zero_values.groupby('season')['orders'].count().reset_index()

# sort data
grouped = grouped.sort_values('orders', ascending=False)

fig = px.bar(grouped, x='season', y='orders', title='Order count by season')
fig.show()

1) Summer has the highest number of orders, followed closely by Spring.
2) Autumn and Winter have a lower count, with Winter having the least.
3) Distribution suggests a seasonal impact on orders. This could be due to several factors, including weather conditions, holiday periods, or consumer behavior that changes with the seasons.


In [306]:
# Map the numeric months to month names
month_names = {1: 'January', 2: 'February', 3: 'March', 4: 'April', 
               5: 'May', 6: 'June', 7: 'July', 8: 'August', 
               9: 'September', 10: 'October', 11: 'November', 12: 'December'}
zero_values['month_name'] = zero_values['month'].map(month_names)

# Group by the new 'month_name' column
grouped = zero_values.groupby('month_name')['orders'].count().reset_index()

# Ensure that the months are ordered correctly
grouped['month_name'] = pd.Categorical(grouped['month_name'], categories=month_names.values(), ordered=True)
grouped = grouped.sort_values('month_name')

# Plot order count by month with month names
fig = px.line(grouped, x='month_name', y='orders', title='Zero Order Count by Month')
fig.show()

C:\Users\monbi\AppData\Local\Temp\ipykernel_2448\4188446003.py:5: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



1) The lowest point occurs in Februay, the month with the fewest occurrences of no orders and the highest peak is in August.
2) Increased variability during the summer months (June to August), with a sharp increase to the highest point in August followed by a sharp decrease in September.


In [307]:
"""
# plot order count by month in a bar chart plotly
grouped = zero_values.groupby('month')['orders'].count().reset_index()

fig = px.line(grouped, x='month', y='orders', title='Zero Order Count by month')
fig.show()
"""

"\n# plot order count by month in a bar chart plotly\ngrouped = zero_values.groupby('month')['orders'].count().reset_index()\n\nfig = px.line(grouped, x='month', y='orders', title='Zero Order Count by month')\nfig.show()\n"

In [308]:
# plot day of the week grouped by season in a grouped bar chart plotly 

grouped = zero_values.groupby(['day_of_week', 'season'])['orders'].count().reset_index()

fig = px.bar(grouped, x='season', y='orders', color='day_of_week', barmode='group' , title='Zero order count by day of week grouped by season')
fig.show()

1) The distribution of zero orders is relatively uniform across different seasons, suggesting that the lack of orders is not strongly influenced by seasonal changes.
2) here is no single day that consistently has the highest or lowest number of zero orders across all seasons, indicating that zero orders are not particularly tied to specific days of the week.
3) While the distributions are similar, there are slight variations in zero order counts between days and seasons, which could be due to natural business cycles or external factors not displayed on the chart.


In [309]:
# Copy relevant columns to a new DataFrame to avoid modifying the original data
df_hm2 = eda_train[['season', 'day_of_week', 'orders']].copy()

# Convert categorical variables to numeric codes
df_hm2['season'] = df_hm2['season'].astype('category').cat.codes
df_hm2['day_of_week'] = df_hm2['day_of_week'].astype('category').cat.codes

# Pivot the data to count zero orders by season and day of the week
pivot_data = df_hm2.pivot_table(index='season', columns='day_of_week', values='orders', aggfunc=lambda x: (x==0).sum())

# Create heatmap
fig = px.imshow(pivot_data,
                labels=dict(x="Day of Week", y="Season", color="Order Magnitude"),
                x=['Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday', 'Sunday'],
                y=['Fall', 'Spring', 'Summer', 'Winter'],
                title='Zero Order Days Heatmap by Day of Week and Season')
fig.show()

# Save the pivot table to df_hm2
df_hm2 = pivot_data


1) There is a noticeable hotspot on Saturday during the Summer, where the color is visibly lighter, indicating this day has a higher frequency of zero orders compared to other days and seasons.
2) Otherwise the heatmap suggests a relatively even distribution of zero-order days across the days of the week and seasons

EDA

    - hourly 

    - shift

    - outliers
Preprocessing

    - outliers

    - data leakage
    
    - scaling

In [310]:
eda_train.head(5)

time city  orders day_of_week  month  season  year  \
10 2021-02-01 10:00:00  BCN      31      Monday      2  Winter  2021   
11 2021-02-01 11:00:00  BCN      29      Monday      2  Winter  2021   
12 2021-02-01 12:00:00  BCN      89      Monday      2  Winter  2021   
13 2021-02-01 13:00:00  BCN     230      Monday      2  Winter  2021   
15 2021-02-01 15:00:00  BCN      84      Monday      2  Winter  2021   

   month_name  
10   February  
11   February  
12   February  
13   February  
15   February

In [312]:
eda_test.head(5)

time city  orders
6480  2021-10-29 18:00:00  BCN      72
1945   2021-04-23 7:00:00  BCN       0
1656   2021-04-11 6:00:00  BCN       3
6658   2021-11-06 4:00:00  BCN       0
3066   2021-06-09 6:00:00  BCN       1


## Modelling

Try different models. Validate each model in a way that would imitate the real problem (every sunday you forecast all of next week). Watch out for data leakage. Evaluate each model on MSE and SMAPE. Which one performs better?


### 3) Random Forest

In [311]:
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_squared_error, mean_absolute_error
from sklearn.preprocessing import OneHotEncoder
from sklearn.compose import ColumnTransformer

# Drop unnecessary columns and prepare data for modeling
columns_to_drop = ['time', 'city', 'month_name']  # Adjust if there are more columns to drop
rf_train = eda_train.drop(columns_to_drop, axis=1, errors='ignore')
rf_test = eda_test.drop(columns_to_drop, axis=1, errors='ignore')

# Define the target and features
y_train = rf_train.pop('orders')
y_test = rf_test.pop('orders')

# Encoding categorical variables
categorical_features = ['day_of_week', 'season']
onehot_encoder = ColumnTransformer(
    transformers=[
        ('cat', OneHotEncoder(), categorical_features)
    ],
    remainder='passthrough'
)

# Apply one-hot encoding to both training and testing data
X_train_encoded = onehot_encoder.fit_transform(rf_train)
X_test_encoded = onehot_encoder.transform(rf_test)

# Random Forest model
rf = RandomForestRegressor(n_estimators=100, random_state=42)
rf.fit(X_train_encoded, y_train)
predictions = rf.predict(X_test_encoded)

# Output predictions
print("Predictions:", predictions)

# Calculate and print evaluation metrics
mse = mean_squared_error(y_test, predictions)
rmse = np.sqrt(mse)
mae = mean_absolute_error(y_test, predictions)
print(f'Random Forest - MSE: {mse}, RMSE: {rmse}, MAE: {mae}')


ValueError: columns are missing: {'year', 'month'}

In [ ]:
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_squared_error, mean_absolute_error
from sklearn.preprocessing import OneHotEncoder
from sklearn.compose import ColumnTransformer

# Assuming 'eda_train' has been used to split into 'X_train', 'X_test', 'y_train', 'y_test'

# Prepare rf_train by excluding 'orders', and unnecessary columns
rf_train = eda_train.drop(['orders', 'time', 'city', 'month_name'], axis=1, errors='ignore')
rf_test = eda_test.drop(['orders', 'time', 'city', 'month_name'], axis=1, errors='ignore')

# Encoding categorical variables
categorical_features = ['day_of_week', 'season']
onehot_encoder = ColumnTransformer(
    transformers=[
        ('cat', OneHotEncoder(), categorical_features)
    ],
    remainder='passthrough'
)

# Apply one-hot encoding to both rf_train and rf_test
X_train_encoded = onehot_encoder.fit_transform(rf_train)
X_test_encoded = onehot_encoder.transform(rf_test)

# Random Forest model
rf = RandomForestRegressor(n_estimators=100, random_state=42)
rf.fit(X_train_encoded, y_train)
predictions = rf.predict(X_test_encoded)

# Output predictions
print("Predictions:", predictions)

# Calculate and print evaluation metrics
mse = mean_squared_error(y_test, predictions)
rmse = np.sqrt(mse)
mae = mean_absolute_error(y_test, predictions)
print(f'Random Forest - MSE: {mse}, RMSE: {rmse}, MAE: {mae}')


In [258]:


# Assuming 'X_train' and 'X_test' already include 'time', 'city', etc., and 'y_train', 'y_test' are already defined

# Prepare features by removing unnecessary columns
columns_to_drop = ['time', 'city', 'month_name']  # Assuming these columns are still in your datasets
X_train.drop(columns_to_drop, axis=1, errors='ignore', inplace=True)
X_test.drop(columns_to_drop, axis=1, errors='ignore', inplace=True)

# Encoding categorical variables
categorical_features = ['day_of_week', 'season']
onehot_encoder = ColumnTransformer(
    transformers=[
        ('cat', OneHotEncoder(), categorical_features)
    ],
    remainder='passthrough'
)

# Apply one-hot encoding to both training and testing data
X_train_encoded = onehot_encoder.fit_transform(X_train)
X_test_encoded = onehot_encoder.transform(X_test)

# Random Forest model
rf = RandomForestRegressor(n_estimators=100, random_state=42)
rf.fit(X_train_encoded, y_train)
predictions = rf.predict(X_test_encoded)

# Output predictions
print("Predictions:", predictions)

# Calculate and print evaluation metrics
mse = mean_squared_error(y_test, predictions)
rmse = np.sqrt(mse)
mae = mean_absolute_error(y_test, predictions)
print(f'Random Forest - MSE: {mse}, RMSE: {rmse}, MAE: {mae}')


ValueError: A given column is not a column of the dataframe

In [222]:
# Assuming 'eda_train' is already defined and contains all the necessary columns
# Prepare rf_train by excluding 'orders', 'time', 'city', and other non-required columns
rf_train = eda_train.drop(['orders', 'time', 'city', 'month_name'], axis=1)

# Encoding categorical variables
categorical_features = ['day_of_week', 'season']
onehot_encoder = ColumnTransformer(
    transformers=[
        ('cat', OneHotEncoder(), categorical_features)
    ],
    remainder='passthrough'
)

# Splitting the dataset into train and test for the Random Forest model
X = rf_train
y = eda_train['orders']  # Target variable
#X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=42)

# Apply one-hot encoding to both training and testing data
X_train_encoded = onehot_encoder.fit_transform(X_train)
X_test_encoded = onehot_encoder.transform(X_test)

# Random Forest model
rf = RandomForestRegressor(n_estimators=100, random_state=42)
rf.fit(X_train_encoded, y_train)
predictions = rf.predict(X_test_encoded)

# Output predictions
print("Predictions:", predictions)

# Calculate and print evaluation metrics
mse = mean_squared_error(y_test, predictions)
rmse = np.sqrt(mse)
mae = mean_absolute_error(y_test, predictions)
print(f'Random Forest - MSE: {mse}, RMSE: {rmse}, MAE: {mae}')

Predictions: [47.47976241 61.22603687 78.02324775 ... 55.79953699 65.04972085
 51.1186252 ]
Random Forest - MSE: 11813.957469724595, RMSE: 108.6920303873499, MAE: 76.75653114380361


In [221]:
# Drop unnecessary columns from X_train and X_test
X_train.drop(['time', 'city'], axis=1, inplace=True)
X_test.drop(['time', 'city'], axis=1, inplace=True)

# Since you mentioned 'day_of_week' and 'season' as categorical features,
# ensure these features are created before encoding if they aren't already included.
# For this example, let's assume 'day_of_week' and 'season' need to be derived.

# Note: You'll need to provide logic or data for these if they're not derived yet.

# Encoding categorical variables
# As 'day_of_week' and 'season' are not listed in the columns, let's assume we need to create these
# Normally, you would extract these from 'time' or another relevant column if applicable.
# Since 'time' is dropped and not used for feature extraction (based on provided columns),
# ensure the required features are derived beforehand or adjust feature selection.

categorical_features = ['hour']  # Temporarily using 'hour' for example, replace with actual categorical features after creation
onehot_encoder = ColumnTransformer(
    transformers=[
        ('cat', OneHotEncoder(), categorical_features)
    ],
    remainder='passthrough'
)

# Apply one-hot encoding to both training and testing data
X_train_encoded = onehot_encoder.fit_transform(X_train)
X_test_encoded = onehot_encoder.transform(X_test)

# Random Forest model
rf = RandomForestRegressor(n_estimators=100, random_state=42)
rf.fit(X_train_encoded, y_train)
predictions = rf.predict(X_test_encoded)

# Output predictions
print("Predictions:", predictions)

# Calculate and print evaluation metrics
mse = mean_squared_error(y_test, predictions)
rmse = np.sqrt(mse)
mae = mean_absolute_error(y_test, predictions)
print(f'Random Forest - MSE: {mse}, RMSE: {rmse}, MAE: {mae}')


KeyError: "['time', 'city'] not found in axis"

In [ ]:
# Convert 'time' into datetime type
data['time'] = pd.to_datetime(data['time'])

# Now that 'time' is confirmed to be datetime, extract 'hour'
data['hour'] = data['time'].dt.hour

# Drop the original 'time', 'city', and 'month_name' columns as they are not needed
data.drop(['time', 'city', 'month_name'], axis=1, inplace=True)

# Encoding categorical variables
categorical_features = ['day_of_week', 'season']
onehot_encoder = ColumnTransformer(
    transformers=[
        ('cat', OneHotEncoder(), categorical_features)
    ],
    remainder='passthrough'
)

# Prepare the features and target
X = data.drop('orders', axis=1)
y = data['orders']
X_encoded = onehot_encoder.fit_transform(X)

# Assume X_train, X_test, y_train, y_test are already defined
# Random Forest model
rf = RandomForestRegressor(n_estimators=100, random_state=42)
rf.fit(X_train, y_train)
predictions = rf.predict(X_test)

# Output predictions
print("Predictions:", predictions)

# Calculate and print evaluation metrics
mse = mean_squared_error(y_test, predictions)
rmse = np.sqrt(mse)
mae = mean_absolute_error(y_test, predictions)
print(f'Random Forest - MSE: {mse}, RMSE: {rmse}, MAE: {mae}')

KeyError: "['month_name'] not found in axis"

In [ ]:
# Convert 'time' into more granular features and remove unwanted columns
data['hour'] = data['time'].dt.hour
data.drop(['time', 'city', 'month_name'], axis=1, inplace=True)  # Drop columns if they are not needed

# Encoding categorical variables
categorical_features = ['day_of_week', 'season']
onehot_encoder = ColumnTransformer(
    transformers=[
        ('cat', OneHotEncoder(), categorical_features)
    ],
    remainder='passthrough'
)

X = data.drop('orders', axis=1)
y = data['orders']

X_encoded = onehot_encoder.fit_transform(X)

# Split data into train and test sets
#X_train, X_test, y_train, y_test = train_test_split(X_encoded, y, test_size=0.25, random_state=42)

# Random Forest model
rf = RandomForestRegressor(n_estimators=100, random_state=42)
rf.fit(X_train, y_train)
predictions = rf.predict(X_test)

# Output predictions
print("Predictions:", predictions)

# Calculate and print evaluation metrics
mse = mean_squared_error(y_test, predictions)
rmse = np.sqrt(mse)
mae = mean_absolute_error(y_test, predictions)
print(f'Random Forest - MSE: {mse}, RMSE: {rmse}, MAE: {mae}')


ValueError: A given column is not a column of the dataframe

Random Forest has best overall performance, this suggests that the time series has strong non-linear components

### 1. Autoregressive Integrated Moving Average (ARIMA) & Seasonal ARIMA (SARIMA)

In [180]:
"""
# Ensure the ARIMA and SARIMA models receive the correct data type
# Convert index to a DatetimeIndex if necessary (if using dates and times)
# This step is crucial if your 'time' column affects the model, and needs DateTime indexing
X_train['time'] = pd.to_datetime(X_train['time'])
X_train.set_index('time', inplace=True)
X_test.set_index('time', inplace=True)

# ARIMA Model
arima_model = ARIMA(y_train, order=(1,1,1))
arima_fitted = arima_model.fit()
arima_forecast = arima_fitted.forecast(steps=len(y_test))

# SARIMA Model
sarima_model = SARIMAX(y_train, order=(1,1,1), seasonal_order=(1,1,1,24))
sarima_fitted = sarima_model.fit()
sarima_forecast = sarima_fitted.forecast(steps=len(y_test))

# Calculate and print errors
for forecast, model_name in [(arima_forecast, 'ARIMA'), (sarima_forecast, 'SARIMA')]:
    mse = mean_squared_error(y_test, forecast)
    rmse = np.sqrt(mse)
    mae = mean_absolute_error(y_test, forecast)
    print(f'{model_name} - MSE: {mse}, RMSE: {rmse}, MAE: {mae}')
"""


"\n# Ensure the ARIMA and SARIMA models receive the correct data type\n# Convert index to a DatetimeIndex if necessary (if using dates and times)\n# This step is crucial if your 'time' column affects the model, and needs DateTime indexing\nX_train['time'] = pd.to_datetime(X_train['time'])\nX_train.set_index('time', inplace=True)\nX_test.set_index('time', inplace=True)\n\n# ARIMA Model\narima_model = ARIMA(y_train, order=(1,1,1))\narima_fitted = arima_model.fit()\narima_forecast = arima_fitted.forecast(steps=len(y_test))\n\n# SARIMA Model\nsarima_model = SARIMAX(y_train, order=(1,1,1), seasonal_order=(1,1,1,24))\nsarima_fitted = sarima_model.fit()\nsarima_forecast = sarima_fitted.forecast(steps=len(y_test))\n\n# Calculate and print errors\nfor forecast, model_name in [(arima_forecast, 'ARIMA'), (sarima_forecast, 'SARIMA')]:\n    mse = mean_squared_error(y_test, forecast)\n    rmse = np.sqrt(mse)\n    mae = mean_absolute_error(y_test, forecast)\n    print(f'{model_name} - MSE: {mse}, RMS